### PR TITLE
MSC3375: Room version 9.

### LINKED ISSUE #3373: join_authorised_via_users_server is not protected from redactions
Room v8 added `join_authorised_via_users_server` to `m.room.member` events to help with authorizing restricted room join events. However, the field isn't protected from redactions, which means the join event will suddenly become invalid (i.e. not pass event auth rules) if it's redacted. At least with synapse, existing servers will still see the member in the room, but any new servers will reject the event and any messages that the user sends.

### PR DESCRIPTION
**Pull request by clokep**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/clokep/room-v9/proposals/3375-room-v9.md)

Fixes #3373

Implementation: https://github.com/matrix-org/synapse/pull/10747

### CODE REVIEW

**Review by turt2live — APPROVED** [review]

File: proposals/3375-room-v9.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+version may include them.
+
+Room version `9` upon being added to the specification shall be considered stable
+and room version `8` shall be considered unstable. No other room versions are
+affected by this MSC.
+
+## Background
+
+The redaction of this key was missing in [MSC3083](https://github.com/matrix-org/matrix-doc/blob/main/proposals/3083-restricted-rooms.md),
+but without redacting it new servers which join a room are unable to properly
```
Comment (richvdh) [inline proposals/3375-room-v9.md:21]: this is backwards. The problem is that it *is* redacted, and we need to protect it therefrom.

File: proposals/3375-room-v9.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+The redaction rules for room version 9 are updated, such that `m.room.member`
+events allow the `join_authorised_via_users_server` key in addition to the
+`membership` key.
+
+Though other MSCs are capable of being included in this version, they do not have
+sufficient implementation to be considered for this room version. A future room
+version may include them.
+
+Room version `9` upon being added to the specification shall be considered stable
+and room version `8` shall be considered unstable. No other room versions are
```
Comment (turt2live) [inline proposals/3375-room-v9.md:15]: We shouldn't touch v8's stability.

https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$Zq0D1HmMjt9bpwiYO8tielwJWENwfEnbNVt8csZQ6g4?via=t2l.io&via=matrix.org&via=libera.chat

Comment (clokep) [reply proposals/3375-room-v9.md:15]: I don't agree. v8 has existed for a short period of time and we should encourage upgrades to v9.

Comment (turt2live) [reply proposals/3375-room-v9.md:15]: We don't have any of the prerequisite support to be able to do that safely or reliably. 

We probably should have thought of a different word to describe room versions, but from #matrix-spec: it's not linked to how buggy a version is but instead whether or not we're planning to break them into not working. There's perfectly functional v8 rooms which don't need upgrades, so our risk surface is minimal.

Comment (clokep) [reply proposals/3375-room-v9.md]: I've updated it, but I do not agree with the rationale. "Unstable" doesn't imply that we're planning to break anything. It means you should not use it.

**Review by uhoreg — APPROVED** [review]

**Review by richvdh — APPROVED** [review]

File: proposals/3375-room-v9.md (hunk 2 above)
Comment (ara4n) [reply proposals/3375-room-v9.md]: (see https://github.com/matrix-org/matrix-doc/issues/3377 for the punchline on this)

**Review by ara4n — APPROVED** [review]

File: proposals/3375-room-v9.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# MSC3375: Room Version 9
+
+A new room version, `9`, is proposed using [room version 8](https://spec.matrix.org/unstable/rooms/v8/)
+as a base and incorporating the following:
+
+The redaction rules for room version 9 are updated, such that `m.room.member`
+events allow the `join_authorised_via_users_server` key in addition to the
+`membership` key.
```
Comment (clokep) [inline proposals/3375-room-v9.md:8]: (Starting a thread just to have a thread.)

@turt2live asked me to do some additional testing with redaction. So I did the following:

1. Created a space and a v9 room as alice.
2. Set the join rules of the room to restricted via the space.
3. Invited bob (from the same server) to the space.
4. Had bob join the room and chat a little.
5. Redact all state events in the room (all `m.room.member` events, `m.room.join_rules`, `m.room.history_visibility`, `m.room.name`, `m.room.guest_access`)
6. Invited charlie (from a new server) to the space
7. Had charlie join the room
8. Ensure that alice, bob, and charlie were all in the room, all able to chat (and see each other's messages).

@turt2live Please let me know if there's additional testing I should do!

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-doc/pull/3387

**Comment by turt2live** [comment]
Merged 🎉

**Pull request merged**